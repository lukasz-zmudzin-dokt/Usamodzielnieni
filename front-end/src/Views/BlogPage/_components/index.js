import NewComment from './NewComment/NewComment';
import CommentsList from './CommentsList/CommentsList';
import BlogContent from './BlogContent/BlogContent';

export {
    NewComment,
    CommentsList,
    BlogContent
}