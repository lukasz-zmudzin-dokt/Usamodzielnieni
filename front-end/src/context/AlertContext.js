import React, { useState, createContext } from "react";

export const AlertContext = createContext({
  open: false,
  message: "",
  setOpen: () => {},
});

export const AlertProvider = (props) => {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState("");
  const [variant, setVariant] = useState("danger") // success
  const data = {
    open,
    message,
    variant,
    changeVisibility: (newOpen) => setOpen(newOpen),
    changeMessage: (newMessage,newVariant = variant) => {
      setOpen(false);
      setMessage(newMessage);
      setVariant(newVariant);
    },
    showAlert:  (newMessage,newVariant) => {
      data.changeMessage(newMessage,newVariant);
      data.changeVisibility(true);
    }
  };
  return <AlertContext.Provider value={data} {...props} />;
};
