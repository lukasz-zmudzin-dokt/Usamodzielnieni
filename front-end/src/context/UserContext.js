import React, { useState } from "react";
import Cookies from "universal-cookie";

export const UserContext = React.createContext({
  type: undefined, // Employer,Standard,Staff
  token: undefined,
  login: () => {},
  logout: () => {}
});
const cookies = new Cookies();

export const UserProvider = props => {
  const [type, setType] = useState(cookies.get("type"));
  const [token, setToken] = useState(cookies.get("token"));
  const user = {
    token,
    type,
    login: (newToken, newType) => {
      cookies.set("token", newToken, { path: "/" });
      cookies.set("type", newType, { path: "/" });
      setToken(newToken);
      setType(newType);
    },
    logout: () => {
      cookies.remove("token", { path: "/" });
      cookies.remove("type", { path: "/" });
      setToken(undefined);
      setType(undefined);
    }
  };
  return <UserContext.Provider value={user} {...props} />;
};
