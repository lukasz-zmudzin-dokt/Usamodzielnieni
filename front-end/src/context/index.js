export { UserContext, UserProvider } from "context/UserContext";
export { AlertContext, AlertProvider } from "context/AlertContext";
export { ChatContext, ChatProvider } from "context/ChatContext";
export {
  NotificationsContext,
  NotificationsProvider,
} from "context/NotificationsContext";
