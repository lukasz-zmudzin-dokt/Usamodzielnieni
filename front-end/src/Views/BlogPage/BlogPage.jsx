import React, { useState, useEffect, useContext } from "react";
import { UserContext, AlertContext } from "context";
import { Container, Card, Alert, CardColumns } from "react-bootstrap";
import { getPosts } from "Views/BlogPage/functions/fetchData";
import BlogPost from "Views/BlogPage/components/SmallBlogPost";
import Filter from "Views/BlogPage/components/Filter";

const BlogPage = () => {
  const [posts, setPosts] = useState([]);
  const [filter, setFilter] = useState({});
  const context = useContext(UserContext);
  const contextA = useContext(AlertContext);
  const [count, setCount] = useState(0);

  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    const loadOffers = async (token) => {
      setIsLoading(true);
      let res;
      try {
        res = await getPosts(token, filter);
      } catch (e) {
        console.log(e);
        res = [];
        contextA.changeMessage("Nie udało się załadować postów");
        contextA.changeVisibility(true);
      }
      setIsLoading(false);
      setCount(res.length);
      setPosts(res);
    };

    loadOffers(context.token);
  }, [context.token, filter]);

  const msg = isLoading ? (
    <Alert variant="info">Ładowanie postów...</Alert>
  ) : (
    count === 0 && (
      <Alert variant="info">Brak postów spełniających podane wymagania.</Alert>
    )
  );

  return (
    <Container>
      <Card>
        <Card.Header as="h2">Blogi</Card.Header>
        <Filter token={context.token} setFilter={setFilter} count={count} />
        {msg ? (
          msg
        ) : (
          <CardColumns className="ml-3 mr-3">
            {posts.map((data) => (
              <BlogPost key={data.id} {...data} />
            ))}
          </CardColumns>
        )}
      </Card>
    </Container>
  );
};

export default BlogPage;
