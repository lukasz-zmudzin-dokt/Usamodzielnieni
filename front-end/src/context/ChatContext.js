import React, {
  createContext,
  useEffect,
  useState,
  useContext,
  useCallback,
} from "react";
import proxy from "config/api";
import { UserContext } from "context";

const getChats = async (token, filters) => {
  let url = `${proxy.chat}?page=${filters.page}&page_size=${filters.pageSize}`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status === 200) {
    return response.json().then((chats) => chats);
  } else {
    throw response.status;
  }
};

export const ChatContext = createContext(null);

export const ChatProvider = (props) => {
  const user = useContext(UserContext);
  const [chats, setChats] = useState([]);
  const [isChatsLoading, setIsChatsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [count, setCount] = useState(0);
  const [filters] = useState({
    page: 1,
    pageSize: 10,
  });
  const [activePage, setActivePage] = useState(1);

  useEffect(() => {
    if (user.token) {
      const loadChats = async (token) => {
        setActivePage(1);
        setError(false);
        setIsChatsLoading(true);
        let loadedChats;
        try {
          loadedChats = await getChats(token, filters);
        } catch (e) {
          console.log(e);
          loadedChats = { count: 0, results: [] };
          setError(true);
        }
        setChats(loadedChats.results);
        setCount(loadedChats.count);
        setIsChatsLoading(false);
      };
      loadChats(user.token);
    }
  }, [filters, user.token]);

  const loadMoreMessages = async () => {
    let res;
    try {
      res = await getChats(user.token, { ...filters, page: activePage + 1 });
      setActivePage(activePage + 1);
    } catch (e) {
      setError(true);
      return;
    }

    setChats([...chats, ...res.results]);
  };

  const loadMessages = async (token) => {
    setActivePage(1);
    let res;
    try {
      res = await getChats(token, filters);
    } catch (e) {
      setError(true);
      return;
    }

    setChats(res.results);
  };

  const data = {
    chats,
    error,
    count,
    loadMoreMessages,
    isChatsLoading,
    loadMessages,
  };

  return <ChatContext.Provider value={data} {...props} />;
};
