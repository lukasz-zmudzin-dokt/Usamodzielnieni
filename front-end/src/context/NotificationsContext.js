import React, {
  useState,
  useContext,
  useEffect,
  useRef,
  createContext,
} from "react";
import { UserContext } from "context";
import proxy from "config/api";
import { paths } from "constants/paths";
import { userTypes } from "constants/userTypes";
import { compile } from "path-to-regexp";

const getNotifications = async (token, type, next) => {
  let url = proxy.notification + "list/all" + (next ? `?page=${next}` : "");
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return response
    .json()
    .then((notifications) => mapNotifications(notifications, type));
};

const getNotificationsCount = async (token) => {
  let url = proxy.notification + "count/unread";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return (await response.json()).unread_count;
};

const markAsRead = async (token) => {
  let url = proxy.notification + "list/all/mark-as-read/";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "POST", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const deleteNotification = async (token, id) => {
  let url = proxy.notification + `delete/${id}/`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "DELETE", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const deleteNotifications = async (token) => {
  let url = proxy.notification + "list/all/delete/";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "DELETE", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const mapNotifications = (notifications, type) => ({
  notifications: notifications.results.map((notification) =>
    mapNotification(notification, type)
  ),
  next: notifications.next?.replace(/.*page=/, "") || undefined,
});

const mapNotification = (notification, type) => ({
  id: notification.slug,
  path: getPath(notification.app, notification.object_id, type),
  title: notification.text,
  time: new Date(notification.timestamp),
  unread: notification.unread,
});

const getPath = (appName, id, type) => {
  const isStandard = type === userTypes.STANDARD;
  if (appName.match(/^cv/)) {
    return isStandard
      ? compile(paths.MY_CVS)({})
      : compile(paths.CV_APPROVAL)({});
  } else if (appName.match(/^jobs/)) {
    return isStandard
      ? compile(paths.JOB_OFFERS)({})
      : compile(paths.MY_OFFERS)({});
  } else if (appName.match(/^account/)) {
    return compile(paths.USER)({});
  } else if (appName.match(/^blog/)) {
    return compile(paths.BLOG_POST)({ id });
  } else {
    return compile(paths.DASHBOARD)({});
  }
};

export const NotificationsContext = createContext(null);

export const NotificationsProvider = (props) => {
  const [notifications, setNotifications] = useState(null);
  const [count, setCount] = useState(null);
  const [next, setNext] = useState();
  const [error, setError] = useState(false);
  const socket = useRef();

  const user = useContext(UserContext);

  useEffect(() => {
    const loadNotifications = async (token, type) => {
      setError(false);
      let values;
      try {
        values = await Promise.all([
          getNotifications(token, type),
          getNotificationsCount(token),
        ]);
      } catch (e) {
        console.log(e);
        setError(true);
        return;
      }
      const [
        { notifications: loadedNotifications, next: loadedNext },
        loadedCount,
      ] = values;
      setCount(loadedCount);
      setNext(loadedNext);
      setNotifications(loadedNotifications);
    };
    if (user.token && user.type) {
      loadNotifications(user.token, user.type);
    }
  }, [user.token, user.type]);

  useEffect(() => {
    if (socket.current) {
      socket.current.close();
      socket.current = null;
    }
    if (user.token && user.type) {
      const url = proxy.wsNotification + "ws";
      try {
        socket.current = new WebSocket(url, user.token);
        // socket.current.onopen = (e) => console.log("onopen", e);
        socket.current.onmessage = (e) => {
          const newNotification = mapNotification(
            JSON.parse(e.data),
            user.type
          );
          setCount((prev) => prev + 1);
          setNotifications((prev) => [newNotification, ...prev]);
        };
        socket.current.onerror = (e) => {
          console.log(e);
          setError(true);
        };
      } catch (e) {
        console.log(e);
      }
    }
  }, [user.token, user.type]);

  const data = {
    notifications,
    count,
    error,
    next,
    socket,
    markAsRead: async () => {
      setCount(0);
      try {
        await markAsRead(user.token);
      } catch (e) {
        setError(true);
        return;
      }
    },
    deleteNotification: async (id) => {
      const notificationToRemove = notifications.find(
        (notification) => notification.id === id
      );
      if (notificationToRemove?.unread && count > 0) {
        setCount((prev) => prev - 1);
      }
      setNotifications((prev) =>
        prev.filter((notification) => notification.id !== id)
      );
      try {
        await deleteNotification(user.token, id);
      } catch (e) {
        setError(true);
        return;
      }
    },
    deleteNotifications: async () => {
      setCount(0);
      setNotifications([]);
      try {
        await deleteNotifications(user.token);
      } catch (e) {
        setError(true);
        return;
      }
    },
    loadMoreNotifications: async () => {
      setNext(null);
      let res;
      try {
        res = await getNotifications(user.token, user.type, next);
      } catch (e) {
        setError(true);
        return;
      }
      setNext(res.next);
      setNotifications((prev) => [...prev, ...res.notifications]);
    },
  };
  return <NotificationsContext.Provider value={data} {...props} />;
};
