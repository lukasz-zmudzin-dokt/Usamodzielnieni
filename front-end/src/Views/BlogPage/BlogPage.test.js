import React from "react";
import { render, waitForElement, fireEvent } from "@testing-library/react";
import BlogPage from "Views/BlogPage";
import { MemoryRouter } from "react-router-dom";
import { AlertContext } from "context";
import proxy from "config/api";

describe("BlogPage", () => {
  let contextA = {
    showAlert: jest.fn(),
  };
  let failFetch;
  let apiFilters = ["abcd"];
  let apiPosts = [
    {
      tags: ["tag"],
      summary: "summary",
      category: "abcd",
      header: "/media/example_img.jpg",
      id: 1,
      title: "tytuł",
    },
    {
      tags: ["tag", "abcd"],
      category: "abcd",
      summary: "summary2",
      header: null,
      id: 2,
      title: "tytuł2",
    },
  ];
  beforeAll(() => {
    global.fetch = jest.fn().mockImplementation((input, init) => {
      return new Promise((resolve, reject) => {
        if (failFetch) {
          resolve({ status: 500 });
        }
        switch (init.method) {
          case "POST":
            resolve({ status: 200 });
            break;
          case "GET":
            if (input.includes(proxy.blog + "blogposts/")) {
              resolve({ status: 200, json: () => Promise.resolve(apiPosts) });
            } else {
              resolve({ status: 200, json: () => Promise.resolve(apiFilters) });
            }

            break;
          default:
            reject({});
            break;
        }
      });
    });
  });

  beforeEach(() => {
    failFetch = false;
    jest.clearAllMocks();
  });

  it("should match snapshot", async () => {
    const { container, getByText } = render(
      <MemoryRouter>
        <BlogPage />
      </MemoryRouter>
    );
    await waitForElement(() => getByText("tytuł"));
    expect(container).toMatchSnapshot();
  });

  it("should load posts", async () => {
    const { getByText } = render(
      <MemoryRouter>
        <BlogPage />
      </MemoryRouter>
    );

    await waitForElement(() => getByText("tytuł"));
    expect(getByText("tytuł")).toBeInTheDocument();
  });

  it("should load filters", async () => {
    const { getAllByText } = render(
      <MemoryRouter>
        <BlogPage />
      </MemoryRouter>
    );

    await waitForElement(() => getAllByText("abcd"));

    expect(getAllByText("abcd").length).toBe(3);
  });

  it("should show message if apiFails", async () => {
    failFetch = true;
    const { getByText } = render(
      <AlertContext.Provider value={contextA}>
        <MemoryRouter>
          <BlogPage />
        </MemoryRouter>
      </AlertContext.Provider>
    );

    await waitForElement(() => getByText("Filtruj posty"));

    expect(contextA.showAlert).toHaveBeenCalledWith(
      "Nie udało się załadować postów"
    );
  });

  describe("Filter", () => {
    it("should be called with appropriate url(2 filters)", async () => {
      const { getByText, getAllByText, getByLabelText } = render(
        <MemoryRouter>
          <BlogPage />
        </MemoryRouter>
      );

      await waitForElement(() => getAllByText("abcd"));

      fireEvent.change(getByLabelText("Kategoria"), {
        target: { value: "abcd" },
      });

      fireEvent.change(getByLabelText("Tag"), {
        target: { value: "abcd" },
      });

      fireEvent.click(getByText("Filtruj posty"));

      await waitForElement(() => getAllByText("abcd"));

      await expect(fetch).toHaveBeenCalledWith(
        `${proxy.blog}blogposts/?category=abcd&tag=abcd`,
        {
          headers: {
            "Content-Type": "application/json",
          },
          method: "GET",
        }
      );
    });

    it("should be called with appropriate url(1 filter - tag)", async () => {
      const { getByText, getAllByText, getByLabelText } = render(
        <MemoryRouter>
          <BlogPage />
        </MemoryRouter>
      );

      await waitForElement(() => getAllByText("abcd"));

      fireEvent.change(getByLabelText("Tag"), {
        target: { value: "abcd" },
      });

      fireEvent.click(getByText("Filtruj posty"));

      await waitForElement(() => getAllByText("abcd"));

      await expect(fetch).toHaveBeenCalledWith(
        `${proxy.blog}blogposts/?tag=abcd`,
        {
          headers: {
            "Content-Type": "application/json",
          },
          method: "GET",
        }
      );
    });

    it("should render post with no tags", async () => {
      apiPosts[0].tags = [];
      const { getByText, getAllByText } = render(
        <MemoryRouter>
          <BlogPage />
        </MemoryRouter>
      );

      await waitForElement(() => getAllByText("abcd"));
      expect(getByText("Brak tagów")).toBeInTheDocument();
    });

    it("should be called with appropriate url(1 filter - category)", async () => {
      const { getByText, getAllByText, getByLabelText } = render(
        <MemoryRouter>
          <BlogPage />
        </MemoryRouter>
      );

      await waitForElement(() => getAllByText("abcd"));

      fireEvent.change(getByLabelText("Kategoria"), {
        target: { value: "abcd" },
      });

      fireEvent.click(getByText("Filtruj posty"));

      await waitForElement(() => getAllByText("abcd"));

      await expect(fetch).toHaveBeenCalledWith(
        `${proxy.blog}blogposts/?category=abcd`,
        {
          headers: {
            "Content-Type": "application/json",
          },
          method: "GET",
        }
      );
    });
  });
});
