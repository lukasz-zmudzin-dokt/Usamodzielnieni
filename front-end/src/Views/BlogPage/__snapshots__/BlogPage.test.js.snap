// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPage should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="card"
    >
      <h2
        class="card-header"
      >
        Blogi
      </h2>
      <form
        class="ml-3 mr-3 mb-3"
      >
        <div
          class="form-row"
        >
          <div
            class="form-group col-md-6 col-12"
          >
            <label
              class="form-label"
              for="category"
            >
              Kategoria
            </label>
            <select
              class="form-control"
              id="category"
            >
              <option
                disabled=""
              >
                -- Wybierz --
              </option>
              <option>
                abcd
              </option>
            </select>
          </div>
          <div
            class="form-group col-md-6 col-12"
          >
            <label
              class="form-label"
              for="tag"
            >
              Tag
            </label>
            <select
              class="form-control"
              id="tag"
            >
              <option
                disabled=""
              >
                -- Wybierz --
              </option>
              <option>
                abcd
              </option>
            </select>
          </div>
        </div>
        <div>
          <button
            class="mr-3 btn btn-primary"
            type="submit"
          >
            Filtruj posty
          </button>
          <button
            class="mr-3 btn btn-outline-primary"
            type="button"
          >
            Wyczyść filtry
          </button>
        </div>
        <div>
          <small
            class="blog__countText"
          >
            Ilość znalezionych postów: 2
          </small>
        </div>
      </form>
      <div
        class="ml-3 mr-3 card-columns"
      >
        <a
          class="sBlogPost__link"
          href="/blog/blogpost/1"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <h3
                class="card-title"
              >
                tytuł
              </h3>
              <p
                class="card-text"
              >
                summary
              </p>
            </div>
            <div
              class="card-footer"
            >
              <small>
                Tagi:
              </small>
              <span
                class="ml-1 badge badge-info"
              >
                tag
              </span>
            </div>
          </div>
        </a>
        <a
          class="sBlogPost__link"
          href="/blog/blogpost/2"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <h3
                class="card-title"
              >
                tytuł2
              </h3>
              <p
                class="card-text"
              >
                summary2
              </p>
            </div>
            <div
              class="card-footer"
            >
              <small>
                Tagi:
              </small>
              <span
                class="ml-1 badge badge-info"
              >
                tag
              </span>
              <span
                class="ml-1 badge badge-info"
              >
                abcd
              </span>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
